### DISCUSSION TITLE
Attiny85: I2C master hangs without slave

### DISCUSSION THREAD

**Discussion by haldark** [SpenceKonde/ATTinyCore#692] [Library discussion]
I have found another bug in the USI I2C code, that makes the I2C bus hang when the addressed slave is not present.

the sketch is very simple. It is used to detect if a specific slave is present on the bus. Essentially the master is just driving the bus with slave address and detecting ack/NACK. Regardless of ack or NACK, The code is supposed to issue stop and leave the bus idle. The error code tells if the targeted slave responded or not.

#include <wire.h>
byte  err;
Wire.begin();
Wire.beginTransmission(0x07); //for slave with ID 7
err = Wire.endTransmission(true):
if (err==0) {

> //slave is found

} else if (err==2) {

> //NACK on slave address. Means slave is not present

} else {

> //unknown error

}

This code works in nano. For nano, the core picks up TWI library, which issues stop even if NACK is received for slave address.

However, same code hangs the bus (no STOP issued) for attiny85 if there is NACK on slave address. The core picks up usi_twi libraries for attiny85. Wire.endTransmission() calls  USI_TWI_Start_Transceiver_With_Data_Stop(). The following piece of code exists (return FALSE) the function without issuing STOP if NACK happens at address phase (line 182).

    /* Clock and verify (N)ACK from slave */
      DDR_USI &= ~(1 << PIN_USI_SDA); // Enable SDA as input.
      if (USI_TWI_Master_Transfer(tempUSISR_1bit) & (1 << TWI_NACK_BIT)) {
        if (USI_TWI_state.addressMode)
          USI_TWI_state.errorState = USI_TWI_NO_ACK_ON_ADDRESS;
        else
          USI_TWI_state.errorState = USI_TWI_NO_ACK_ON_DATA;
        return (FALSE);
      }
      USI_TWI_state.addressMode = FALSE; // Only perform address transmission once.
    }

Code for issuing STOP is at line 210, which is not executed in NACK situation. These 3 lines should be also included right before return(FALSE) in the code above.

  if (stop) {
    USI_TWI_Master_Stop(); // Send a STOP condition on the TWI bus.
  }

With the bus at this state, other slaves on the bus act weird.

**What I am trying to do**
In my setup, at85 is I2C master with 2 slaves on I2C bus. One is a mangnetometer, an the other is a nano acting as slave (for debugging purpose) with address 7. At85 periodically reads all 3 axes of the magnetometer, and immediately writes same values to slave 7. Then it reads a register in slave 7 for further actions. This goes on and on. The nano prints all received values in human readable format to serial port. It also accepts commands from serial port. When a command comes via serial port, nano stores that in a variable. When at85 reads the register in slave 7, this variable is returned to at85. At85 takes further action based on this variable value, like configuring the magnetometer, storing/retrieving parameters in at85 eeprom, etc.

basically, nano is being used as a debugging channel, as it is very hard to debug using just at85.

the setup works beautifully when nano is present on the bus. But I want to burn the code unchanged into at85 even for production version, where debug-nano will be absent. The unchanged code allows me to debug even on production board just by connecting the nano on the I2C bus. The idea is that on production board, after at85 taken mag reading, it will first detect if the slave 7 is present. If present, it will relay the mag value to slave 7, otherwise skip this step.

the setup works when master is implemented using a nano. But fails if it’s at85. That’s because of the above bug, which leaves the bus hanging, and magnetometer does not behave as expected for subsequent readings.

**Comment by SpenceKonde** [comment]
How was this solved on the '328p? 

I find the I2C protocol very confounding, like it was created for the explicit purpose of confusing people trying to debug it,

**Reply by haldark** [reply]
For 328p, TWDR is defined, hence twi.c is picked up by your Wire.cpp/h. Wire.endTransmission calls twi_writeTo.

328p has HW I2C, and fires interrupt every time TWDR transmission is completed. The ISR checks the error status. Here is what happens ( line 427 of twi.c) when Slave address experiences NACK.


case TW_MT_SLA_NACK:  // address sent, nack received
      twi_error = TW_MT_SLA_NACK;
      twi_stop();
      break;

That twi_stop() saves the day.

**Comment by hmeijdam** [comment]
Remarkable. I am using a AT85 with a sketch where I scan for the address of a LCD 16x2 I2C display first, before I activate it, as the display backpack prints come with different I2C addresses. Also when no display is connected to the I2C bus it will detect that.
The only way it will hang is if no external pull-up resistors are connected to the I2C bus. Activating the internal pull-ups does not help, as they seem to get deactivated by wire.

I extracted below part of my sketch, that will demonstrate scanning of the I2C display address and then printing the address. 
If no display found the led will blink faster.
I use the display for debugging.

[edit] maybe important to notice. I am using Attinycore 2.0 and the digispark Liquidcrystal_I2C library for the Attiny and in that library I had to replace "tinywirem" back into "wire"  as the Attinycore now supports the standard wire command

```
/*************************************************************************************************
  Pinlayout of the ATTINY 85
                                          +-------+
              (PCINT5/RESET/ADC0/dW) PB5 -| 1   8 |- VCC
 PIN 3 (PCINT3/XTAL1/CLKI/OC1B/ADC3) PB3 -| 2   7 |- PB2 (SCK/USCK/SCL/ADC1/T0/INT0/PCINT2) PIN 2
 PIN 4 (PCINT4/XTAL2/CLKO/OC1B/ADC2) PB4 -| 3   6 |- PB1 (MISO/DO/AIN1/OC0B/OC1A/PCINT1) PIN 1 PWM
                                     GND -| 4   5 |- PB0 (MOSI/DI/SDA/AIN0/OC0A/OC1A/AREF/PCINT0) PIN 0 PWM
                                          +-------+

 *************************************************************************************************/
#define LED1       PB1  // DEBUG LED (via resistor) 

#include <LiquidCrystal_I2C.h>       // for LCD w/ GPIO MODIFIED for the ATtiny85
LiquidCrystal_I2C * lcd;             // create a class pointer to avoid having to specify I2C address here

int main(void) {
  DDRB |= _BV(LED1);//Led1 as output

  // local variables:
  uint8_t lcd_found = 0;

  Wire.begin(); // start I2C interface

  for (uint8_t lcd_addr = 0x3f; lcd_addr >= 0x20; lcd_addr--)
    /*The PCF8574 based backpacks are in two address ranges
      0x20 to 0x27
      0x38 to 0x3f*/
  {
    Wire.beginTransmission (lcd_addr); // scan the I2C bus
    if (Wire.endTransmission () == 0) {// found something
      lcd = new LiquidCrystal_I2C (lcd_addr, 16, 2); // use found I2C address for LCD display
      lcd->init();
      lcd->backlight();
      lcd->setCursor(0, 0);
      lcd->print("LCD display at");
      lcd->setCursor(0, 1);
      lcd->print(lcd_addr, HEX);
      delay(2000);
      lcd->clear();
      lcd_found = 1;
      break;
    }
  }

  while (1) { // loop
    if (lcd_found) {
      lcd->setCursor(0, 0);
      lcd->print("Hello World");
      delay(1000);
      lcd->clear();
      delay(1000);
      PORTB ^= _BV(LED1); //toggle led
    }
    else //lcd not found
      PORTB ^= _BV(LED1); //toggle led
      delay(500);
  }
}
```

**Reply by haldark** [reply]
I know for fact that tinywirem does not enable pull-ups for I2C master. You must use external pull-ups when using tinywirem. I have been in than route, before staring with Wire library with attinycore. In fact when I switched to Wire, my setup did not work as I still had ext pull-ups on my setup and USI_TWI also enabled pull-up. It worked only after I removed ext pull-ups.

**Comment by haldark** [comment]
@hmeijdam , I will give it a try with external pull-ups, but it does not make sense to me. The setup works perfectly with one slave (magnetometer), which means at85 pull-ups are active and doing their job. The setup even works when both magnetometer (slave ID 0xE) and debug-nano (slave ID 0x7) are connected. The setup also can read magnetometer when debug-nano is not present. It hangs only when it addresses debug-nano and gets NACK.

**Reply by hmeijdam** [reply]
Yes, one slave is good enough. No need to add extra pullups. I only need to add the extra pullups if the SDA and SCL are totally unconnected to anything.

**Comment by SpenceKonde** [comment]
Why are you using a modified library and tinywirem when ATTinyCore includes a universal wire that works with unmodified libraries?
